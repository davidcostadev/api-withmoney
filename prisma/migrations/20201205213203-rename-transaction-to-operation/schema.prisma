// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url = "***"
}

generator client {
  provider = "prisma-client-js"
}

model Account {
  id        String      @id @default(uuid())
  name      String
  userId    String
  createdAt DateTime    @default(now())
  updatedAt DateTime    @default(now())
  deletedAt DateTime?
  user      User        @relation(fields: [userId], references: [id])
  operations Operation[]
}

model Category {
  id        String        @id @default(uuid())
  userId    String
  name      String
  type      TransactionType
  createdAt DateTime      @default(now())
  updatedAt DateTime      @default(now())
  deletedAt DateTime?
  user      User          @relation(fields: [userId], references: [id])
  operations Operation[]
}

model Operation {
  id         String        @id @default(uuid())
  accountId  String
  categoryId String?
  userId     String
  name       String
  value      Float
  isPaid     Boolean       @default(false)
  type       TransactionType
  createdAt  DateTime      @default(now())
  updatedAt  DateTime      @default(now())
  deletedAt  DateTime?
  account    Account       @relation(fields: [accountId], references: [id])
  category   Category?     @relation(fields: [categoryId], references: [id])
  user       User          @relation(fields: [userId], references: [id])
}

model User {
  id                   String      @id @default(cuid())
  email                String      @unique
  password             String
  firstName            String
  lastName             String
  hasVerifiedEmail     Boolean     @default(false)
  hashToVerifyEmail    String?     @unique
  hashToChangePassword String?     @unique
  birthDay             DateTime?
  createdAt            DateTime    @default(now())
  updatedAt            DateTime    @default(now())
  deletedAt            DateTime?
  accounts             Account[]
  categories           Category[]
  operations            Operation[]
}

enum TransactionType {
  CreditCard
  Deposit
  FixedExpense
  VariableExpense
}
