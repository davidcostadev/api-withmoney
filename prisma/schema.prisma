generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Account {
  id         String      @id @default(uuid())
  name       String
  userId     String
  createdAt  DateTime    @default(now())
  updatedAt  DateTime    @default(now()) @updatedAt
  deletedAt  DateTime?
  user       User        @relation(fields: [userId], references: [id])
  operations Operation[]
}

model Category {
  id        String          @id @default(uuid())
  userId    String
  name      String
  type      TransactionType
  createdAt DateTime        @default(now())
  updatedAt DateTime        @default(now()) @updatedAt
  deletedAt DateTime?
  User      User            @relation(fields: [userId], references: [id])
  Operation Operation[]
}

model Operation {
  id         String          @id @default(uuid())
  accountId  String
  categoryId String?
  userId     String
  name       String
  value      Float
  isPaid     Boolean         @default(false)
  type       TransactionType
  createdAt  DateTime        @default(now())
  updatedAt  DateTime        @default(now()) @updatedAt
  deletedAt  DateTime?
  Account    Account         @relation(fields: [accountId], references: [id])
  Category   Category?       @relation(fields: [categoryId], references: [id])
  User       User            @relation(fields: [userId], references: [id])
}

model User {
  id                   String      @id @default(uuid())
  email                String      @unique
  password             String
  firstName            String
  lastName             String
  hasVerifiedEmail     Boolean     @default(false)
  hashToVerifyEmail    String?     @unique
  hashToChangePassword String?     @unique
  birthDay             DateTime?
  createdAt            DateTime    @default(now())
  updatedAt            DateTime    @default(now()) @updatedAt
  deletedAt            DateTime?
  accounts             Account[]
  categories           Category[]
  Operation            Operation[]
}

enum TransactionType {
  CreditCard
  Deposit
  FixedExpense
  VariableExpense
}
