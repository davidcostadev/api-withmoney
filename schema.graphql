### This file was generated by Nexus Schema
### Do not make changes to this file directly


type Account {
  createdAt: DateTime!
  deletedAt: DateTime
  id: String!
  name: String!
  operations(after: OperationWhereUniqueInput, before: OperationWhereUniqueInput, first: Int, last: Int, orderBy: [OperationOrderByInput!], where: OperationWhereInput): [Operation!]!
  updatedAt: DateTime!
  user: User!
  userId: String!
}

input AccountCreateInput {
  name: String!
}

input AccountListRelationFilter {
  every: AccountWhereInput
  none: AccountWhereInput
  some: AccountWhereInput
}

input AccountOrderByInput {
  createdAt: SortOrder
  deletedAt: SortOrder
  id: SortOrder
  name: SortOrder
  updatedAt: SortOrder
  userId: SortOrder
}

input AccountUpdateInput {
  name: String!
}

input AccountWhereInput {
  AND: [AccountWhereInput!]
  createdAt: DateTimeFilter
  deletedAt: DateTimeNullableFilter
  id: StringFilter
  name: StringFilter
  NOT: [AccountWhereInput!]
  operations: OperationListRelationFilter
  OR: [AccountWhereInput!]
  updatedAt: DateTimeFilter
  user: UserWhereInput
  userId: StringFilter
}

input AccountWhereUniqueInput {
  id: String
}

type AuthPayload {
  token: String
  user: User
}

input BoolFilter {
  equals: Boolean
  not: NestedBoolFilter
}

type Category {
  createdAt: DateTime!
  deletedAt: DateTime
  id: String!
  name: String!
  operations(after: OperationWhereUniqueInput, before: OperationWhereUniqueInput, first: Int, last: Int, orderBy: [OperationOrderByInput!], where: OperationWhereInput): [Operation!]!
  type: TransactionType!
  updatedAt: DateTime!
  user: User!
}

input CategoryCreateInput {
  name: String!
  type: TransactionType!
}

input CategoryListRelationFilter {
  every: CategoryWhereInput
  none: CategoryWhereInput
  some: CategoryWhereInput
}

input CategoryOrderByInput {
  createdAt: SortOrder
  deletedAt: SortOrder
  id: SortOrder
  name: SortOrder
  type: SortOrder
  updatedAt: SortOrder
  userId: SortOrder
}

input CategoryUpdateInput {
  name: String!
  type: TransactionType!
}

input CategoryWhereInput {
  AND: [CategoryWhereInput!]
  createdAt: DateTimeFilter
  deletedAt: DateTimeNullableFilter
  id: StringFilter
  name: StringFilter
  NOT: [CategoryWhereInput!]
  operations: OperationListRelationFilter
  OR: [CategoryWhereInput!]
  type: EnumTransactionTypeFilter
  updatedAt: DateTimeFilter
  user: UserWhereInput
  userId: StringFilter
}

input CategoryWhereUniqueInput {
  id: String
}

"""
A date string, such as 2007-12-03, compliant with the `full-date` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar Date

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

input DateTimeFilter {
  equals: DateTime
  gt: DateTime
  gte: DateTime
  in: [DateTime!]
  lt: DateTime
  lte: DateTime
  not: NestedDateTimeFilter
  notIn: [DateTime!]
}

input DateTimeNullableFilter {
  equals: DateTime
  gt: DateTime
  gte: DateTime
  in: [DateTime!]
  lt: DateTime
  lte: DateTime
  not: NestedDateTimeNullableFilter
  notIn: [DateTime!]
}

input EnumTransactionTypeFilter {
  equals: TransactionType
  in: [TransactionType!]
  not: NestedEnumTransactionTypeFilter
  notIn: [TransactionType!]
}

input FloatFilter {
  equals: Float
  gt: Float
  gte: Float
  in: [Float!]
  lt: Float
  lte: Float
  not: NestedFloatFilter
  notIn: [Float!]
}

type Mutation {
  changePassword(hash: String!, password: String!): String
  checkHashEmail(hash: String!): String
  createOneAccount(data: AccountCreateInput!): Account!
  createOneCategory(data: CategoryCreateInput!): Category!
  createOneOperation(data: OperationCreateInput!): Operation!
  deleteOneOperation(where: OperationWhereUniqueInput!): Operation
  login(email: String!, password: String!): AuthPayload
  register(user: RegisterInput!): String
  requestChangePassword(email: String!): String
  restoreOneOperation(where: OperationWhereUniqueInput!): Operation
  updateOneAccount(data: AccountUpdateInput!, where: AccountWhereUniqueInput!): Account!
  updateOneCategory(data: CategoryUpdateInput!, where: CategoryWhereUniqueInput!): Category!
  updateOneOperation(data: OperationUpdateInput!, where: OperationWhereUniqueInput!): Operation!
  updateOneUser(user: UserUpdateInput!): User!
}

input NestedBoolFilter {
  equals: Boolean
  not: NestedBoolFilter
}

input NestedDateTimeFilter {
  equals: DateTime
  gt: DateTime
  gte: DateTime
  in: [DateTime!]
  lt: DateTime
  lte: DateTime
  not: NestedDateTimeFilter
  notIn: [DateTime!]
}

input NestedDateTimeNullableFilter {
  equals: DateTime
  gt: DateTime
  gte: DateTime
  in: [DateTime!]
  lt: DateTime
  lte: DateTime
  not: NestedDateTimeNullableFilter
  notIn: [DateTime!]
}

input NestedEnumTransactionTypeFilter {
  equals: TransactionType
  in: [TransactionType!]
  not: NestedEnumTransactionTypeFilter
  notIn: [TransactionType!]
}

input NestedFloatFilter {
  equals: Float
  gt: Float
  gte: Float
  in: [Float!]
  lt: Float
  lte: Float
  not: NestedFloatFilter
  notIn: [Float!]
}

input NestedStringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: NestedStringFilter
  notIn: [String!]
  startsWith: String
}

input NestedStringNullableFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: NestedStringNullableFilter
  notIn: [String!]
  startsWith: String
}

type Operation {
  account: Account!
  category: Category
  createdAt: DateTime!
  deletedAt: DateTime
  id: String!
  name: String!
  type: TransactionType!
  updatedAt: DateTime!
  user: User!
  value: Float!
}

input OperationCreateInput {
  accountId: String!
  categoryId: String
  name: String!
  type: TransactionType!
  value: Float!
}

input OperationListRelationFilter {
  every: OperationWhereInput
  none: OperationWhereInput
  some: OperationWhereInput
}

input OperationOrderByInput {
  accountId: SortOrder
  categoryId: SortOrder
  createdAt: SortOrder
  deletedAt: SortOrder
  id: SortOrder
  isPaid: SortOrder
  name: SortOrder
  type: SortOrder
  updatedAt: SortOrder
  userId: SortOrder
  value: SortOrder
}

input OperationUpdateInput {
  accountId: String!
  categoryId: String
  name: String!
  type: TransactionType!
  value: Float!
}

input OperationWhereInput {
  account: AccountWhereInput
  accountId: StringFilter
  AND: [OperationWhereInput!]
  category: CategoryWhereInput
  categoryId: StringNullableFilter
  createdAt: DateTimeFilter
  deletedAt: DateTimeNullableFilter
  id: StringFilter
  isPaid: BoolFilter
  name: StringFilter
  NOT: [OperationWhereInput!]
  OR: [OperationWhereInput!]
  type: EnumTransactionTypeFilter
  updatedAt: DateTimeFilter
  user: UserWhereInput
  userId: StringFilter
  value: FloatFilter
}

input OperationWhereUniqueInput {
  id: String
}

type Query {
  findUniqueAccount(where: AccountWhereUniqueInput): Account
  me: User
}

enum QueryMode {
  default
  insensitive
}

input RegisterInput {
  birthday: Date
  email: String!
  firstName: String!
  lastName: String!
  password: String!
}

enum SortOrder {
  asc
  desc
}

input StringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  mode: QueryMode
  not: NestedStringFilter
  notIn: [String!]
  startsWith: String
}

input StringNullableFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  mode: QueryMode
  not: NestedStringNullableFilter
  notIn: [String!]
  startsWith: String
}

type Subscription {
  userSignedIn(userId: String!): User
  userUpdated(userId: String!): User
}

enum TransactionType {
  CreditCard
  Deposit
  FixedExpense
  VariableExpense
}

"""The `Upload` scalar type represents a file upload."""
scalar Upload

type User {
  accounts(after: AccountWhereUniqueInput, before: AccountWhereUniqueInput, first: Int, last: Int, orderBy: [AccountOrderByInput!], where: AccountWhereInput): [Account!]!
  birthDay: DateTime
  categories(after: CategoryWhereUniqueInput, before: CategoryWhereUniqueInput, first: Int, last: Int, orderBy: [CategoryOrderByInput!], where: CategoryWhereInput): [Category!]!
  createdAt: DateTime!
  deletedAt: DateTime
  email: String!
  firstName: String!
  hasVerifiedEmail: Boolean!
  id: String!
  lastName: String!
  operations(after: OperationWhereUniqueInput, before: OperationWhereUniqueInput, first: Int, last: Int, orderBy: [OperationOrderByInput!], where: OperationWhereInput): [Operation!]!
  updatedAt: DateTime!
}

input UserUpdateInput {
  birthday: Date
  email: String
  name: String
  nickname: String
  phone: String
}

input UserWhereInput {
  accounts: AccountListRelationFilter
  AND: [UserWhereInput!]
  birthDay: DateTimeNullableFilter
  categories: CategoryListRelationFilter
  createdAt: DateTimeFilter
  deletedAt: DateTimeNullableFilter
  email: StringFilter
  firstName: StringFilter
  hashToChangePassword: StringNullableFilter
  hashToVerifyEmail: StringNullableFilter
  hasVerifiedEmail: BoolFilter
  id: StringFilter
  lastName: StringFilter
  NOT: [UserWhereInput!]
  operations: OperationListRelationFilter
  OR: [UserWhereInput!]
  password: StringFilter
  updatedAt: DateTimeFilter
}
