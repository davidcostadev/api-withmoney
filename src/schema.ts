import * as path from 'path';
import * as types from './types';

import { makeSchema } from '@nexus/schema';
import { nexusSchemaPrisma } from 'nexus-plugin-prisma/schema';

export const schema = makeSchema({
  types,
  plugins: [
    nexusSchemaPrisma({
      // Generate typefiles on any occasions
      shouldGenerateArtifacts: true,
      outputs: {
        typegen: __dirname + '/generated/typegen-nexus-plugin-prisma.d.ts',
      },
    }),
  ],
  outputs: {
    schema: path.join(__dirname, './../schema.graphql'),
    typegen: path.join(__dirname, './generated/nexus.ts'),
  },
});
