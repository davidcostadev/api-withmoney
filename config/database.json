{
  "development": {
    "username": "root",
    "password": "123",
    "database": "mymoney_dev",
    "host": "127.0.0.1",
    "dialect": "mysql",
    "logging": false
  },
  "test": {
    "username": "root",
    "password": "123",
    "database": "mymoney_test",
    "host": "127.0.0.1",
    "dialect": "mysql",
    "logging": false
  },
  "production": {
    "use_env_variable": "DATABASE_URL"
  }
}
