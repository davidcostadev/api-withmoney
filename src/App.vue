<template>
  <div id="app">
    <md-toolbar>
      <h3 class="md-title">With Money</h3>
    </md-toolbar>
    <md-table>
      <md-table-row>
        <md-table-head md-numeric>ID</md-table-head>
        <md-table-head>Nome</md-table-head>
        <md-table-head>Valor</md-table-head>
        <md-table-head>Pago?</md-table-head>
        <md-table-head>Data</md-table-head>
      </md-table-row>
    </md-table>
    <md-content>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit minus doloremque, animi culpa dolores explicabo vel voluptatum. Iure sed rem, veritatis maiores adipisci quam tempore? Magni perferendis ab fugiat tempora?
    </md-content>
    <router-view/>
  </div>
</template>

<script>
export default {
  name: 'App',
};
</script>

<style>
#app {

}
  .md-content {
    padding: 16px;
  }
</style>
