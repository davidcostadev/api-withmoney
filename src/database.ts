import mongoose, { Document } from 'mongoose';

const { MONGODB } = process.env;

if (!MONGODB) {
  throw new Error('MONGODB env is not defined');
}

mongoose.connect(MONGODB, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
});

const userSchema = new mongoose.Schema({
  firstName: { type: String },
  lastName: { type: String },
  email: { type: String },
  password: { type: String },
});
export interface IUser extends Document {
  password: string;
  email: string;
  firstName: string;
  lastName: string;
}

export const Users = mongoose.model<IUser>('Users', userSchema);
