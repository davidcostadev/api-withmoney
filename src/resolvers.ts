import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { Users } from './database';

if (!process.env.SECRET_SALT) {
  /* istanbul ignore next */
  throw new Error('SECRET_SALT env is undefined');
}

if (!process.env.SECRET_KEY) {
  /* istanbul ignore next */
  throw new Error('SECRET_KEY env is undefined');
}

const { SECRET_SALT, SECRET_KEY } = process.env;


interface IRegister {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}

interface ILogin {
  email: string;
  password: string;
}

export const resolvers = {
  Query: {
    users: () => {
      return Users.find();
    },
  },
  Mutation: {
    register: async (root: any, data: IRegister) => {
      const searchUser = await Users.findOne({
        email: data.email,
      });

      if (searchUser) {
        throw new Error('The User already exists!');
      }

      const newPassword = await bcrypt.hash(data.password, parseInt(SECRET_SALT, 10));

      await Users.create({
        ...data,
        password: newPassword,
        hasVerifiedEmail: false,
      });

      return 'OK';
    },

    login: async (root: any, data: ILogin) => {
      const user = await Users.findOne({
        email: data.email,
      });

      if (!user) {
        throw new Error('Email or Password invalid');
      }

      if (await !bcrypt.compare(data.password, user.password)) {
        throw new Error('Email or Password invalid');
      }

      return {
        token: jwt.sign({
          id: user.id,
          firstName: user.firstName,
          lastName: user.lastName,
          email: user.email,
          hasVerifiedEmail: user.hasVerifiedEmail
        }, SECRET_KEY),
      };
    },
  },
};
