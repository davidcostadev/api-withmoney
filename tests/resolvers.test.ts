import { ApolloServer, gql } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
import { typeDefs } from '../src/schema';
import { resolvers } from '../src/resolvers';
import { Users } from '../src/database';

const entity = {
  id: '5e584b5994515952cab4e97b',
  email: 'xexeu@gmail.com',
  password: '$2b$04$V2BrnJds.QlkvQQaGgW1.u7xbP4RDKz729JXh/yxCdrI4fY/ROdfq',
  firstName: 'Jhonny',
  lastName: 'Michel',
  hasVerifiedEmail: false,
};

jest.mock('../src/database', () => ({
  Users: {
    findOne: jest.fn(),
    create: jest.fn(),
  },
}));

jest.mock('nodemailer', () => ({
  createTransport: () => ({
    sendMail: jest.fn().mockResolvedValue(null),
  }),
}));

it('Should return a OK on register mutation', async () => {
  Users.findOne = jest.fn().mockResolvedValue(null);
  Users.create = jest.fn().mockResolvedValue(entity);

  const server = new ApolloServer({ typeDefs, resolvers });

  const { mutate } = createTestClient(server);

  const REGISTER_USER = gql`
    mutation Register(
      $firstName: String!
      $lastName: String!
      $email: String!
      $password: String!
    ) {
      register(firstName: $firstName, lastName: $lastName, email: $email, password: $password)
    }
  `;

  const res = await mutate({
    mutation: REGISTER_USER,
    variables: {
      firstName: 'Jhonny',
      lastName: 'Michel',
      email: 'xexeu@gmail.com',
      password: '123456',
    },
  });

  expect(res).toMatchSnapshot();
});

it('Should return a token on login mutation', async () => {
  Users.findOne = jest.fn().mockResolvedValue(entity);
  const server = new ApolloServer({ typeDefs, resolvers });

  const { mutate } = createTestClient(server);

  const LOGIN_USER = gql`
    mutation Login($email: String!, $password: String!) {
      login(email: $email, password: $password) {
        token
      }
    }
  `;

  const res = await mutate({
    mutation: LOGIN_USER,
    variables: {
      email: 'xexeu@gmail.com',
      password: '123456',
    },
  });

  expect(res).toEqual({
    data: {
      login: {
        token: expect.any(String),
      },
    },
    errors: undefined,
    extensions: undefined,
    http: expect.any(Object),
  });
});

it('Should return a OK on checkHashEmail mutation', async () => {
  Users.findOne = jest.fn().mockResolvedValue({
    firstName: 'David',
    email: 'davidcostadev@gmail.com',
    save: () => Promise.resolve(),
  });

  const server = new ApolloServer({ typeDefs, resolvers });

  const { mutate } = createTestClient(server);

  const CHECK_HASH_EMAIL = gql`
    mutation checkHashEmail($hash: String!) {
      checkHashEmail(hash: $hash)
    }
  `;

  const res = await mutate({
    mutation: CHECK_HASH_EMAIL,
    variables: {
      hash: '12345',
    },
  });

  expect(res).toMatchSnapshot();
});
