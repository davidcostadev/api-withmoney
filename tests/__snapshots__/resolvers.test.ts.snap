// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return a OK on register mutation 1`] = `
Object {
  "data": Object {
    "register": "OK",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
